File: exercise.js
// Problem 2
/*
Rephrase Question: Find the halfway point of a list, or the two items in the middle if there isn't a halfway point

Pseudocode:
    - Create Node Class
    - Create a Linked List Class
    - Use the find method to find middle of the list 
    - If list contains an even number of nodes, take both the head and head.next nodes in the middle
	
*/

// Solution:
var middleNode = function(head){
    let fast = slow = head
    while (fast&& fast.next) {
        fast = fast.next.next
        slow = slow.next
    }
    return slow, slow.next
}

// Problem 4

/*
Rephrase Question: Find the halfway point of a list, or the two items in the middle if there isn't a halfway point

Pseudocode:
    - Create a Map
    - Have a start and end point
    - Return 0 if the string is empty or no letter
    - Set both pointers to be at the beginning of string
        - Have the max substring length = 1
    -While end pointer is less than N:
        - Add the next character in the map and move end pointer to the right
        - If the letter is found again, remove the leftmost character and update max length
	
*/

// Solution

var lengthOfLongestSubstringKDistinct = function(s, k) {
    if (!s || k === 0) {
        return 0;
    }
    let map = new Map(); 
    let start = 0;
    let end = 0;
    let r = 0;
    while (end <= s.length) {
        if (map.size === k && !map.has(s[end]) || end === s.length) {
            r = Math.max(r, end - start);
            while (map.size === k) {
                const count = map.get(s[start]);
                if (count - 1 === 0) {
                    map.delete(s[start]);
                } else {
                    map.set(s[start], count - 1);
                }
                start++;
            }
        } 
        if (map.has(s[end])) {
            map.set(s[end], map.get(s[end]) + 1);
        } else {
            map.set(s[end], 1);
        }
        end++;
    }
    return r;  
};

File: exercise.py
# Problem 1
'''
list1 = [1,2,3,4,5,5,6,7,8,9,]
return True

list2 = [9,4,2,3,4,8,10,11]
return True

list3 = [2,3,4,5,6,7,8,9,0]
return false

Pseudo:
Function of find duplicates( takes in array)
	empty array start
	for number in the array:
		if the number id in the array:
			return True
		else:
			append number to empty array
	return False
'''

#Actual: Simplest yet not most efficient
def find_dup(array):
    dup_array = []
    for number in array:
        if number in array:
            return True
        else:
            dup_array.append(number)
    #Returning false if no dup is found
    return False



# Other Way: Problem 1
'''
Pseudo:
function(array):
	empty dictionary
	for loop over the array
'''
#Code
def find_dup(array):
	dup_dict = {}
	for number in array:
		if dup_dict 
    #Returning false if no dup is found
	return False

# Problem 4
'''
Restate the problem:
Given a string of text and the number k, find the number of words that appear multiple 
times in the tex, up to the number k in the number of words. Return them in an array of 
decreasing order.
ex. I have, I will, I want. It is that I will have the item. I want it. 
k = 3
answer = ["I","have","will"]

Pseudocode:
	- Essentially make a dictogram function that takes in a string
	- Converts to a list, strip spaces and punctuation
	- Go through list and add to dictionary
		- Check if in dictionary
			- Add one to the count value of the dictioanry
		- If not
			- Add to dictionary with count 1
Organize the dictionary by value amounts - hopefully in descending order
Add the key values to a list via k
return list
'''

# Code

def frequent_words(string):
	word_dict = {}
	some_list = []
	return_list = []
	words_list = [line.strip() for line in string]
	if words_list is not None:
		for word in words_list:
			if word not in word_dict:
				word_dict[word] = 1
			else:
				word_dict[word] += 1
			sorted_dict = sorted(word_dict.items(), key = lambda item: item[1], reverse = True)

			for object in sorted_dict:
				some_list.append(object)
			for i range(0..k):
				return_list.append(some_list[i])

	return return_list

# Problem 2
'''
Rephrase Question: Find the halfway point of a list, or the two items in the middle if there isn't a halfway point

Pseudocode:
	- Create Node Class
	
'''